erdos/utils.py

```diff
@@ -1,11 +1,49 @@
 import logging
+import time
 from functools import wraps
+from threading import Thread, Lock
 
 _freq_called = set([])
 
 
+def deadline(*expected_args):
+    """
+    Deadline decorator to be used for restraining computation latency
+    :param duration: [int/float] computation's duration constrain in ms
+    :param deadline_missing_callback: [str] function name to call when the deadline is missed
+    :return:
+    """
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+
+            def check_deadline(target_ending_time):
+                global task_finished
+                while not task_finished and time.time() < target_ending_time:
```

Comment:
The code here is not thread safe because you're not locking `task_finished`. I checked and it looks like there's a more elegant way to do this using Conditions in Python. The code would look like this:
```
condition = threading.Condition()


def check_deadline(end_time):
    condition.acquire()
    condition.wait(timeout=end_time - time.time())
    condition.release()
    if time.time() >= end_time:
        getattr(args[0], expected_args[1])()

check_thread = threading.Thread(target=check_deadline,
                                args=(target_ending_time,))
check_thread.start()
condition.acquire()
func(*args, **kwargs)
condition.notify()
condition.release()
check_thread.join()
```